crates/control/src/ball_filter.rs

```diff
@@ -125,34 +132,52 @@ impl BallFilter {
                 )
             });
 
-            for ball in balls {
-                let mean_position = ball.percept_in_ground.mean.framed().as_point();
-                let is_hypothesis_detected = |hypothesis: &BallHypothesis| {
-                    distance(
-                        hypothesis
-                            .choose_ball(filter_parameters.resting_ball_velocity_threshold)
-                            .position,
-                        mean_position,
-                    ) < filter_parameters.measurement_matching_distance
-                };
-                let is_any_hypothesis_updated = self.ball_filter.update(
+            let match_matrix =
+                mahalanobis_matrix_of_hypotheses_and_percepts(&self.ball_filter.hypotheses, &balls);
+
+            let assignment = AssignmentProblem::from_costs(match_matrix).solve();
+
+            let mut used_percepts = vec![];
+
+            for (hypothesis, assigned_percept) in self
+                .ball_filter
+                .hypotheses
+                .iter_mut()
+                .zip_eq(assignment.iter())
+            {
+                if let Some(assigned_percept) = assigned_percept {
+                    let mahalanobis_distance = -assigned_percept.cost;
+                    if mahalanobis_distance > filter_parameters.maximum_matching_cost {
+                        hypothesis.validity *= 1.0 / 7.0;
+                        continue;
+                    }
+                    let validity_increase = assigned_percept.cost.exp();
+                    let percept = balls[assigned_percept.to];
+                    used_percepts.push(assigned_percept.to);
+                    hypothesis.update(detection_time, percept.percept_in_ground, validity_increase);
+                }
+            }
+
+            let unused_percepts = {
+                let mut all_percepts = balls.clone();
+                used_percepts.sort_unstable();
+                for index in used_percepts.into_iter().rev() {
+                    all_percepts.remove(index);
+                }
+                all_percepts
+            };
+
+            for percept in unused_percepts {
+                self.ball_filter.spawn(
                     detection_time,
-                    ball.percept_in_ground,
-                    is_hypothesis_detected,
+                    percept.percept_in_ground,
+                    Matrix4::from_diagonal(&filter_parameters.noise.initial_covariance),
                 );
-                if !is_any_hypothesis_updated {
-                    self.ball_filter.spawn(
-                        detection_time,
-                        mean_position,
-                        Matrix4::from_diagonal(&filter_parameters.noise.initial_covariance),
-                        Matrix2::from_diagonal(&filter_parameters.noise.initial_covariance.xy()),
-                    )
-                }
             }
         }
 
         let is_hypothesis_valid = |hypothesis: &BallHypothesis| {
```

Comment:
For me this inplace definition of the filtering criterion was kinda confusing at first. Maybe we could refactor these into separate functions?  Same with `should_merge_hypothesis`

Reply:
discussed, not needed